controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -380,20 +480,35 @@ public void getEpochSegments(GetEpochSegmentsRequest request, StreamObserver<Seg
 
     @Override
     public void getSegments(GetSegmentsRequest request, StreamObserver<SegmentsAtTime> responseObserver) {
-        log.debug("getSegments called for stream " + request.getStreamInfo().getScope() + "/" +
-                request.getStreamInfo().getStream());
-        authenticateExecuteAndProcessResults(() -> this.grpcAuthHelper.checkAuthorizationAndCreateToken(
-                authorizationResource.ofStreamInScope(request.getStreamInfo().getScope(),
-                        request.getStreamInfo().getStream()),
-                AuthHandler.Permissions.READ_UPDATE),
-                delegationToken -> {
-                    logIfEmpty(delegationToken, "getSegments", request.getStreamInfo().getScope(),
-                            request.getStreamInfo().getStream());
-                    return controllerService.getSegmentsAtHead(request.getStreamInfo().getScope(),
-                            request.getStreamInfo().getStream())
+        final StreamInfo streamInfo = request.getStreamInfo();
+        final String scope = streamInfo.getScope();
+        final String stream = streamInfo.getStream();
+        log.debug("getSegments called for stream " + scope + "/" + stream);
+
+        final boolean isDelegationTokenRequested =
+                streamInfo.getRequestedPermission().equals(AccessOperation.NONE.toString()) ? false : true;
```

Comment:
wouldnt requested permission for older clients be null?
should we invert this check to avoid NPE?

Reply:
Yes, thanks for spotting the bug. I've made the change now.